library/Fission/Storage/IPFS/Add.hs

```diff
@@ -0,0 +1,167 @@
+  -> IPFS.Name
+  -> m (Either IPFS.Error.Add IPFS.SparseTree)
+addFile raw name =
+  IPFS.Proc.run opts raw >>= \case
+    (ExitSuccess, result, _) ->
+      IPFS.Pin.add (mkCID <| UTF8.textShow result) >>= \case
+        Left err ->
+          return . Left . UnknownAddErr <| UTF8.textShow err
+
+        Right _ ->
+          case CL.lines result of
+            [inner, outer] ->
+              let
+                sparseTree  = Directory [(Hash rootCID, fileWrapper)]
+                fileWrapper = Directory [(fileName, Content fileCID)]
+                rootCID     = CID <| UTF8.textShow outer
+                fileCID     = CID . UTF8.stripN 1 <| UTF8.textShow inner
+                fileName    = Key name
+              in
+                return <| Right sparseTree
+
+            bad ->
+              return . Left . UnexpectedOutput <| UTF8.textShow bad
+
+
+    (ExitFailure _, _, err) ->
+      return . Left . UnknownAddErr <| UTF8.textShow err
+
+    where
+      opts = [ "add"
+             , "-wq"
+             , "--stdin-name"
+             , unName name
+             ]
+
+addPath ::
+  ( RIOProc           cfg m
+  , Has IPFS.Timeout  cfg
+  , Has IPFS.BinPath  cfg
+  )
+  => FilePath
+  -> m (Either IPFS.Error.Add CID)
+addPath path = IPFS.Proc.run ["add", "-HQ", path] "" >>= pure . \case
+    (ExitSuccess, result, _) ->
+      case CL.lines result of
+        [cid] -> Right . mkCID . UTF8.stripN 1 <| UTF8.textShow cid
+        bad   -> Left . UnexpectedOutput <| UTF8.textShow bad
+
+    (ExitFailure _, _, err) ->
+      Left . UnknownAddErr <| UTF8.textShow err
+
+addDir ::
+  ( MonadRIO cfg m
+  , HasProcessContext cfg
+  , HasLogFunc cfg
+  , Has IPFS.BinPath cfg
+  , Has IPFS.Timeout cfg
+  )
+  => [Glob.Pattern]
+  -> FilePath
+  -> m (Either IPFS.Error.Add IPFS.CID)
+addDir ignored path = doesFileExist path >>= \case
+  True -> addPath path
+  False -> walkDir ignored path
+
+walkDir ::
+  ( MonadRIO cfg m
+  , HasProcessContext cfg
+  , HasLogFunc cfg
+  , Has IPFS.BinPath cfg
+  , Has IPFS.Timeout cfg
+  )
+  => [Glob.Pattern]
+  -> FilePath
+  -> m (Either IPFS.Error.Add IPFS.CID)
+walkDir ignored path = do
+  files <- listDirectory path
+  let toAdd = removeIgnored ignored files
+  results <- mapM (addDir ignored . combine path) toAdd 
+  case lefts results of
+    [] -> do
+      let names = IPFS.Name <$> toAdd
+      links <- zipWithM DAG.Link.create (rights results) names
+      case lefts links of
+        [] -> DAG.putNode Node
+              { dataBlock = "CAE="
+              , links = rights links
+              }
+        _  -> return . Left <| firstErr links
+    _  -> return . Left <| firstErr results
+
+removeIgnored :: [Glob.Pattern] -> [FilePath] -> [FilePath]
+removeIgnored ignored files = List.filter (not . matchesAny ignored) files
+
+matchesAny :: [Glob.Pattern] -> FilePath -> Bool
+matchesAny globs path = List.any (\x -> Glob.match x path) globs
+              
+firstErr :: [Either a b] -> a
```

Comment:
Thoughts on this? Given a list of potentially multiple errors, we break and return the first in the list

Reply:
On a mechanical note, won't this fail at runtime if there are no `Left`s?

Reply:
Have you considered using something like `foldM` instead of `mapM` above? If you want to fail on the first `Left`, that may be a way to do it rather than deconstructing a mix of `Either`s to see if there's ant `Left`s

Reply:
In the only place I'm using it, I check to make sure that some `lefts` is not empty. But I shouldn't rely on a check like that to keep this from blowing up.

I reimplemented this using `foldM`. I think it definitely makes it easier to grok. Let me know what you think! 

Is there a way to fail and exit `foldM` after encountering your first `Left`? Right now I"m just pattern matching and returning a `Left`.